ibm/resource_ibm_pi_volume_attach.go

```diff
@@ -191,32 +158,66 @@ func resourceIBMPIVolumeAttachDelete(ctx context.Context, d *schema.ResourceData
 	return nil
 }
 
-func isWaitForIBMPIVolumeAttachAvailable(ctx context.Context, client *st.IBMPIVolumeClient, id, powerinstanceid string, timeout time.Duration) (interface{}, error) {
+func isWaitForIBMPIVolumeAttachAvailable(ctx context.Context, client *st.IBMPIVolumeClient, id, cloudInstanceID, pvmInstanceID string, timeout time.Duration) (interface{}, error) {
 	log.Printf("Waiting for Volume (%s) to be available for attachment", id)
 
 	stateConf := &resource.StateChangeConf{
 		Pending:    []string{"retry", helpers.PIVolumeProvisioning},
 		Target:     []string{helpers.PIVolumeAllowableAttachStatus},
-		Refresh:    isIBMPIVolumeAttachRefreshFunc(client, id, powerinstanceid),
+		Refresh:    isIBMPIVolumeAttachRefreshFunc(client, id, cloudInstanceID, pvmInstanceID),
 		Delay:      10 * time.Second,
-		MinTimeout: 2 * time.Minute,
-		Timeout:    10 * time.Minute,
+		MinTimeout: 30 * time.Second,
+		Timeout:    timeout,
 	}
 
 	return stateConf.WaitForStateContext(ctx)
 }
 
-func isIBMPIVolumeAttachRefreshFunc(client *st.IBMPIVolumeClient, id, powerinstanceid string) resource.StateRefreshFunc {
+func isIBMPIVolumeAttachRefreshFunc(client *st.IBMPIVolumeClient, id, cloudInstanceID, pvmInstanceID string) resource.StateRefreshFunc {
 	return func() (interface{}, string, error) {
 		vol, err := client.Get(id)
 		if err != nil {
 			return nil, "", err
 		}
 
-		if vol.State == "in-use" {
+		if vol.State == "in-use" && StringContains(vol.PvmInstanceIds, pvmInstanceID) {
 			return vol, helpers.PIVolumeAllowableAttachStatus, nil
 		}
 
 		return vol, helpers.PIVolumeProvisioning, nil
 	}
 }
+
+func isWaitForIBMPIVolumeDetach(ctx context.Context, client *st.IBMPIVolumeClient, id, cloudInstanceID, pvmInstanceID string, timeout time.Duration) (interface{}, error) {
+	log.Printf("Waiting for Volume (%s) to be available after detachment", id)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"detaching", helpers.PowerVolumeAttachDeleting},
+		Target:     []string{helpers.PIVolumeProvisioningDone},
+		Refresh:    isIBMPIVolumeDetachRefreshFunc(client, id, cloudInstanceID, pvmInstanceID),
+		Delay:      10 * time.Second,
+		MinTimeout: 30 * time.Second,
+		Timeout:    timeout,
+	}
+
+	return stateConf.WaitForStateContext(ctx)
+}
+
+func isIBMPIVolumeDetachRefreshFunc(client *st.IBMPIVolumeClient, id, cloudInstanceID, pvmInstanceID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		vol, err := client.Get(id)
+		if err != nil {
+			return nil, "", err
```

Comment:
Could you please check for the error code 404 and `return vol, helpers.PIVolumeProvisioningDone, nil` which indicates that the volume is deleted and hence attachment does not exist.
This will help against unwanted race condition with simultaneous destroy of `ibm_pi_volume_attach` and `ibm_pi_volume`.

Reply:
Can we type check the error instead? 404 can be part of some other error where some id can contain 404 in its value.

Reply:
added type check for 404